deps/buffer.lua

```diff
@@ -207,9 +243,18 @@ Buffer.writeInt32LE = Buffer.writeUInt32LE
 Buffer.writeInt32BE = Buffer.writeUInt32BE
 
 function Buffer:toString(i, j)
-  local offset = i and i - 1 or 0
-  if (offset < 0 or offset > self.length) or (j and j > self.length) then error("Range out of bounds") end
-  return ffi.string(self.ctype + offset, (j or self.length) - offset)
+  i = i and i - 1 or 0
+  j = j or self.length
+  if (i < 0 or i > self.length) or (j and j > self.length) or (i > j) then
+    error("Range out of bounds")
+  end
+  if not hasFfi then
+    if self.length <= 0 then
+      return ''
+    end
+    return char(unpack(self.ctype, i, j - 1))
```

Comment:
using unpack here is unsound, in Luajit and Lua 5.1 it will fail for any buffer of size > MAXCSTACK (by default 8000)

Reply:
The unpack usage here is meant as an optimization, this way we only have to call `string.char` once.

Reply:
Here are the benchmarks for `table.unpack` vs `table.concat`:
```
Buffer library benchmarks using Lua tables (JIT ON, table.unpack)
Creating 10e5 16-byte buffers: 0.291
Creating 10e5 128-byte buffers: 0.767
Creating 10e5 16-byte buffers with content: 0.326
Creating 10e5 128-byte buffers with content: 0.838
Creating and calling toString on 10e5 16-byte buffers: 0.442
Creating and calling toString on 10e5 128-byte buffers: 1.256
Write operations with 10e6 iteration: 0.0030000000000001
Read operations with 10e8 iteration: 0.241
```
```
Buffer library benchmarks using Lua tables (JIT ON, table.concat)
Creating 10e5 16-byte buffers: 0.362
Creating 10e5 128-byte buffers: 0.87
Creating 10e5 16-byte buffers with content: 0.31
Creating 10e5 128-byte buffers with content: 0.858
Creating and calling toString on 10e5 16-byte buffers: 0.818
Creating and calling toString on 10e5 128-byte buffers: 3.477
Write operations with 10e6 iteration: 0.0030000000000001
Read operations with 10e8 iteration: 0.243
```
and the implementation is
```lua
  if not has_ffi then
    if self.length <= 0 then
      return ''
    end
    local tbl = {}
    for c = i, j - 1 do
      tbl[#tbl+1] = char(self.ctype[c])
    end
    return table.concat(tbl)
  end
```

Seems like using unpack is worth the performance gain here, what do you think?

Reply:
I don't doubt that passing unpack directly into string.char is fast. I said it is unsound, it *will* fail for buffer sizes >8KB and may fail earlier for strange configurations (anything that reduces MAXCSTACK).

So either this needs to be an optimization with a fallback, or we drop the performance gain here entirely in favor of having code that works.